File: app/controllers/checkin_controller.rb
# frozen_string_literal: true

class CheckinController < ApplicationController
  before_action :require_login

  def new
    flash.clear
    @reasons = Checkin.reasons
  end

  def create
    flash.clear
    @user = Student.find_by_id(session[:current_user_id])
    @checkin = Checkin.new
    @checkin.update(time: Time.current, student_id: @user.id, reason: params[:checkin][:reason])
    if @checkin.save
      redirect_to root_path, flash: { success: "Success! You've been checked in!" }
    else
      redirect_to root_path, flash: { error: 'Something went wrong, please try again' }
    end
  end

  private

  def require_login
    redirect_to root_path, flash: { error: 'Please log-in first!' } unless
      session.key?(:current_user_id) && Student.find_by_id(session[:current_user_id])
  end
end

File: app(react)/src/controllers/checkin_controller.js
class CheckinController extends ApplicationController {
    constructor(...args) {
        super(...args);
    }
    new() {
        flash.clear;
        this._reasons = Checkin.reasons
      };
      
    create() {
        flash.clear;
        this._user = Student.find_by_id(session.current_user_id);
        this._checkin = new Checkin;
      
        this._checkin.update({
          time: Time.current,
          student_id: this._user.id,
          reason: params.checkin.reason
        });
      
        if (this._checkin.save) {
          redirect_to(
            root_path,
            {flash: {success: "Success! You've been checked in!"}}
          )
        } else {
          redirect_to(
            root_path,
            {flash: {error: "Something went wrong, please try again"}}
          )
        }
      };
      
    private;
      
    require_login() {
        if (!session.key("current_user_id") || !Student.find_by_id(session.current_user_id)) {
          redirect_to(root_path, {flash: {error: "Please log-in first!"}})
        }
      }
}